ipa-core/src/helpers/buffers/circular.rs

```diff
@@ -0,0 +1,809 @@
+/// is to be able to distinguish empty and full buffers.
+///
+/// This implementation does not perform checks in optimized builds,
+/// relying on [`OrderingSender`] to enforce correctness. If taken away,
+/// necessary adjustments need to be made to avoid data corruption.
+///
+/// ## Alternative implementations
+/// If alignment to `read_size` is too much, a [`BipBuffer`] can be used instead.
+///
+/// ## Future improvements
+/// [`OrderingSender`] currently synchronizes reader and writers, but it does not
+/// have to if this implementation is made thread-safe. There exists a well-known
+/// lock-free FIFO implementation for a single producer, single consumer that uses
+/// atomics for read and write pointers. We can't make use of it as is because there
+/// are more than one writer. However, [`OrderingSender`] already knows how to allow
+/// only one write at a time, so it could be possible to make the entire
+/// implementation lock-free.
+///
+/// [`BipBuffer`]: <https://www.codeproject.com/Articles/3479/The-Bip-Buffer-The-Circular-Buffer-with-a-Twist>
+/// [`OrderingSender`]: crate::helpers::buffers::OrderingSender
+/// [`can_read`]: CircularBuf::can_read
+/// [`close`]: CircularBuf::close
+pub struct CircularBuf {
+    /// write pointer, points to the beginning of the next write slice
+    write: usize,
+    /// read pointer, points to the beginning of the next read slice
+    read: usize,
+    /// The size of read returned from [`take`] if buffer is not closed and not empty.
+    read_size: usize,
+    /// The size of one write
+    write_size: usize,
+    /// Whether this buffer is closed
+    closed: bool,
+    /// Actual data, stored inside a contiguous region in memory.
+    data: Vec<u8>,
+}
+
+impl CircularBuf {
+    /// Constructs a new instance of [`CircularBuf`] with reserved `capacity` bytes and specified
+    /// `write_size` and `read_size` bytes.
+    ///
+    /// ## Panics
+    /// If any of the following conditions are met:
+    /// * Any provided value is 0
+    /// * `write_size` is not a multiple of `capacity`
+    /// * `read_size` is not a multiple of `write_size`
+    /// * `read_size` is smaller than `write_size`
+    /// * `read_size` is larger than `capacity`
+    pub fn new(capacity: usize, write_size: usize, read_size: usize) -> Self {
+        debug_assert!(capacity > 0 && write_size > 0 && read_size > 0,
+                      "Capacity \"{capacity}\", write \"{write_size}\" and read size \"{read_size}\" must all be greater than zero"); // enforced at the level above, so debug_assert is fine
+        debug_assert!(
+            capacity % write_size == 0,
+            "\"{write_size}\" write size must divide capacity \"{capacity}\""
+        );
+        debug_assert!(
+            read_size % write_size == 0,
+            "\"{write_size}\" write size must divide read_size \"{read_size}\""
+        );
+        Self {
+            write: 0,
+            read: 0,
```

Comment:
I don't understand what is going on with coverage, but I double checked that `len` code and other places are covered